Modix.RemoraShim/Responders/MuteRoleConfigurationResponder.cs

```diff
@@ -0,0 +1,164 @@
+            _authorizationContextService = authorizationContextService;
+            _channelApi = channelApi;
+            _designatedChannelService = designatedChannelService;
+            _moderationService = moderationService;
+        }
+
+        static MuteRoleConfigurationResponder()
+            => _muteDeniedPermissionSet = new(_muteDeniedPermissions);
+
+        public Task HandleNotificationAsync(Discord.ReadyNotification notification, CancellationToken cancellationToken)
+        {
+            _readySignal.Set();
+            return Task.CompletedTask;
+        }
+
+        public async Task<Result> RespondAsync(IGuildCreate gatewayEvent, CancellationToken ct = default)
+        {
+            _readySignal.Wait(ct);
+
+            try
+            {
+                var authenticationResult = await _authorizationContextService.SetCurrentAuthenticatedUserAsync(gatewayEvent.ID);
+                if (!authenticationResult.IsSuccess)
+                    return authenticationResult;
+
+                var muteRole = await _moderationService.GetOrCreateDesignatedMuteRoleAsync(gatewayEvent.ID.Value, _authorizationService.CurrentUserId!.Value);
+
+                var unmoderatedChannels = await _designatedChannelService.GetDesignatedChannelIdsAsync(gatewayEvent.ID.Value, DesignatedChannelType.Unmoderated);
+
+                var textChannels = gatewayEvent.Channels.Value
+                    .Where(x => x.Type is ChannelType.GuildText or ChannelType.GuildVoice)
+                    .Where(x => !unmoderatedChannels.Contains(x.ID.Value))
+                    .ToImmutableArray();
+
+                var errorChannels = new List<IChannel>();
+
+                foreach (var channel in textChannels)
+                {
+                    var permissions = channel.PermissionOverwrites.Value
+                        .SingleOrDefault(x => x.ID.Value == muteRole.Id);
+
+                    if (permissions is null || _muteDeniedPermissions.Any(x => !permissions.Deny.HasPermission(x)))
+                    {
+                        var editChannelResult = await _channelApi.EditChannelPermissionsAsync(channel.ID, new Snowflake(muteRole.Id), deny: _muteDeniedPermissionSet,
+                            type: PermissionOverwriteType.Role, reason: "Denying thread permissions for the mute role.", ct: ct);
+                        if (!editChannelResult.IsSuccess)
+                            errorChannels.Add(channel);
+                    }
+                }
+
+                return errorChannels.Count > 0
+                    ? Result.FromError(new ChannelConfigurationError("mute role configuration", errorChannels))
+                    : Result.FromSuccess();
+            }
+            catch (Exception ex)
+            {
+                return Result.FromError(new ExceptionError(ex));
+            }
+        }
+
+        public async Task<Result> RespondAsync(IChannelCreate gatewayEvent, CancellationToken ct = default)
```

Comment:
should these be tied to the IThreadCreate/IThreadUpdate, etc responders? I'm also not seeing where these are limited to threads?

Reply:
is that what `if (gatewayEvent.Type is not (ChannelType.GuildText or ChannelType.GuildVoice))` is doing?

Reply:
Nope, threads don't have permissions on their own; they inherit their permissions from the channel they're associated with.

If you have a use public/private thread permission, you can still send messages in a thread even if you don't have the send messages permission. So, the purpose of this class is to deny the use public/private threads permissions to muted users so that muting will work properly.